lib/rucio/core/transfer.py

```diff
@@ -1071,6 +1071,44 @@ def apply(self, ctx: RequestRankingContext, source: RequestSource) -> "Optional[
         return int(len(path) > 1)
 
 
+class FailureRate(SourceRankingStrategy):
+    """
+    A source ranking strategy that ranks source nodes based on their failure rates for the past hour. Failure rate is
+    calculated by dividing files failed by files attempted.
+    """
+    class _FailureRateStat:
+        def __init__(self):
```

Comment:
You are still missing some type annotations (the return values of the two __init__ . The pipeline is too strinct to my liking on this subject, but we have to live with it :)  https://github.com/rucio/rucio/actions/runs/7096607130/job/19322721542